### ISSUE TITLE
The content-type of the AtmosphereRequest should be kept empty when the content is empty

### ISSUE THREAD

**Issue by elakito** [Atmosphere/atmosphere#2141]
Currently, when the content-type is not set in its Builder instance when creating a AtmosphereRequest but the parent request is set, the content-type value is taken from its parent request message. But this logic will likely interfere with the operation matching (because of the unexpected content-type value) when the content of the request is empty and that was the reason why the content-type was not set.

**Comment by celsosantarosa** [comment]
This behaviour is causing a http code 415 with Spring RestEasy (RESTEASY003065: Cannot consume content type). For some reason, the builder body is always empty even, though I have sent a valid request. The same request works with atmosphere 2.4.3.

I am using the following command to make a request to my API: curl -H "Content-Type: application/json" -X POST -d '{"name":"xyz"}' http://localhost:8080/...

**Comment by elakito** [comment]
@celsosantarosa Do you have a test case that I can look at? I don't know what exactly in happening in your scenario. Your curl request is a plain http POST request and that should be recognized with the body content. The subsequent request over the connection has no body nor content-type and this patch will leave the content-type empty.

**Comment by vjoomens** [comment]
Hi, I have the same issue. I have a jax-rs application using rest easy underneath atmosphere using an atmosphere handler. The patch clears the content-type, so rest easy is unable to consume the request.

<atmosphere-handlers>
	
	<atmosphere-handler support-session="false" context-root="/*" class-name="org.atmosphere.handler.ReflectorServletProcessor">
		<property name="servletClassName" value="org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher" />
	</atmosphere-handler>
</atmosphere-handlers>

**Comment by elakito** [comment]
@vjoomens What is the signature of the operation? Is it a GET or a POST with an empty content? It looks like this is specific to RestEasy that fails to handle an empty content message when the content-type is omitted. We can provide an option to turn back on the old behavior. But I would like to know the exact circumstance.

**Comment by vjoomens** [comment]
@elaktito, thank you. Everything works fine with version 2.4.3. It seems that the content type header is cleared always when passing the request to rest easy. 
I think the best solution would be to not clear the content type header when the request is passed on to an atmosphere handler configured in the atmosphere. Xml file.

**Comment by elakito** [comment]
@vjoomens for jersey we needed to clear the content-type if the call is to be matched against the operation that doesn't expect any content. As I mentioned before, the content-type is not propagated only if the body is empty. Which version of RestEasy are you using? I saw this ticket https://issues.jboss.org/browse/RESTEASY-549 which mentions that rest-easy's empty content-type handling issue. So, I was wondering if you have this problem with an older RestEasy version or the current version still has this behavior.
thanks.
regards, aki

**Comment by celsosantarosa** [comment]
@elakito I don't think you understand the issue. I took a look at the code and it seems like you are not setting the body attribute in the AtmosphereRequest.Builder. If you take a look at the method public final static AtmosphereRequest wrap(HttpServletRequest request) in the AtmosphereRequestImpl class, you will see that you are calling the build method(the last line of it) without set the body to the builder. So, if you look at the public AtmosphereRequest build() method, you will see again that you are always setting the body to NULL_BODY cause the body attribute is always null. 

I hope It might help you to fix the issue without add a new configuration option that, IMHO, it is unnecessary.

**Comment by elakito** [comment]
we revert this change as this issue needs to be fixed differently.

**Comment by elakito** [comment]
@celsosantarosa thanks for pointing out the issue. I reverted both changes, so that the content-type is passed to the AtmosphereRequest when instantiated by its wrap method. To solve the original issue, I changed Builder to allow the content-type to be reset. This change does not add any processing overhead (as opposed to if setting the body by instantiating the body).

If this has solved your issue, we can cut a release (2.4.6) with this change.
regards, aki
